runtime-tests/tests/project_registration.rs

```diff
@@ -21,127 +21,261 @@
-        .get_org(register_org.org_id.clone())
+// Verify that a same project can not be re-registered under the same user or org.
+#[async_std::test]
+async fn re_register_project_same_domain_entity() {
+    let (client, _) = Client::new_emulator();
+    let (author, user_id) = key_pair_with_associated_user(&client).await;
+
+    for domain in project_domains(&client, &author, &user_id).await {
+        let checkpoint_id = submit_ok(
+            &client,
+            &author,
+            message::CreateCheckpoint {
+                project_hash: H256::random(),
+                previous_checkpoint_id: None,
+            },
+        )
         .await
-        .unwrap()
+        .result
         .unwrap();
-    assert_eq!(org.projects.len(), 1);
-    assert!(
-        org.projects.contains(&project.name.clone()),
-        "Org does not contain the added project."
-    );
-
-    assert_eq!(
-        client.free_balance(&org.account_id).await.unwrap(),
-        initial_balance - random_fee,
-        "The tx fee was not charged properly."
-    );
+
+        let message = random_register_project_message(&domain, checkpoint_id);
+        submit_ok(&client, &author, message.clone()).await;
+
+        // Duplicate submission with a different metadata.
+        let registration_2 = submit_ok(
+            &client,
+            &author,
+            message::RegisterProject {
+                metadata: Bytes128::random(),
+                ..message.clone()
+            },
+        )
+        .await;
+
+        assert_eq!(
+            registration_2.result,
+            Err(RegistryError::DuplicateProjectId.into())
+        );
+
+        let project = client
+            .get_project(message.project_name, message.project_domain)
+            .await
+            .unwrap()
+            .unwrap();
+        // Assert that the project data was not altered during the
+        // attempt to re-register the already existing project.
+        assert_eq!(message.metadata, project.metadata);
+
+        let projects_list = match &domain {
+            ProjectDomain::Org(org_id) => {
+                let org = client.get_org(org_id.clone()).await.unwrap().unwrap();
+                org.projects
+            }
+            ProjectDomain::User(user_id) => {
+                let user = client.get_user(user_id.clone()).await.unwrap().unwrap();
+                user.projects
+            }
+        };
+
+        // Assert that the number of projects in the involved domain didn't change.
+        assert_eq!(projects_list.len(), 1);
+        assert!(
+            projects_list.contains(&project.name),
+            format!(
+                "Registered project not found in the project list of {}",
+                domain
+            )
+        );
+    }
 }
 
+// Verify that two different orgs can have a project identified by the same name.
 #[async_std::test]
-async fn register_project_under_inexistent_org() {
+async fn re_register_project_org_domain() {
```

Comment:
This name does not reflect what the test does. How about
```suggestion
async fn register_same_project_name_under_different_orgs() {
```

Reply:
hm maybe it's not a clear name, but it does reflect what it does: it registered the same project under the `org` domain. But yeah, it's not clear, I like your suggestion better.